### ISSUE TITLE
pyenv tries to overwrite global Python for Python 3.6.2

### ISSUE THREAD

**Issue by danieljaouen** [pyenv/pyenv#1003]
```pyenv``` tries to overwrite my global ```/Applications/Python Launcher.app``` file.  Is there a way to install this file to ```~/Applications``` instead of ```/Applications```?:

```
test -d "/Applications/Python 3.6/Python Launcher.app" && rm -r "/Applications/Python 3.6/Python Launcher.app"
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Info.plist? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/MacOS/Python Launcher? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/MacOS? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/Credits.rtf? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MainMenu.nib/classes.nib? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MainMenu.nib/info.nib? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MainMenu.nib/objects.nib? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MainMenu.nib? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MyDocument.nib/classes.nib? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MyDocument.nib/info.nib? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MyDocument.nib/objects.nib? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MyDocument.nib? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/classes.nib? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/info.nib? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/objects.nib? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/factorySettings.plist? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonCompiled.icns? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonLauncher.icns? n
override rw-r--r--  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonSource.icns? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents/Resources? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app/Contents? n
override rwxr-xr-x  root/wheel for /Applications/Python 3.6/Python Launcher.app? n
/bin/cp -r "Python Launcher.app" "/Applications/Python 3.6"
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Info.plist: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/MacOS/Python Launcher: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/Credits.rtf: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MainMenu.nib/classes.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MainMenu.nib/info.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MainMenu.nib/objects.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MyDocument.nib/classes.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MyDocument.nib/info.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/MyDocument.nib/objects.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/classes.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/info.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/objects.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/factorySettings.plist: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonCompiled.icns: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonLauncher.icns: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonSource.icns: Permission denied
make[2]: *** [install] Error 1
make[1]: *** [install_PythonLauncher] Error 2
make: *** [frameworkinstallapps] Error 2

BUILD FAILED (OS X 10.12.6 using python-build 1.1.4-2-gf9183b5)

Inspect or clean up the working tree at /var/folders/fm/z9mbpztn17776d_msnfzh84w0000gp/T/python-build.20171005161847.51606
Results logged to /var/folders/fm/z9mbpztn17776d_msnfzh84w0000gp/T/python-build.20171005161847.51606.log

Last 10 log lines:
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/classes.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/info.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/English.lproj/PreferenceWindow.nib/objects.nib: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/factorySettings.plist: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonCompiled.icns: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonLauncher.icns: Permission denied
cp: /Applications/Python 3.6/Python Launcher.app/Contents/Resources/PythonSource.icns: Permission denied
```

**Comment by techdragon** [comment]
Im getting this as well with the current latest version of pyenv and a 3.6.3 framework build.

**Comment by techdragon** [comment]
It also seems that pyenv does not respect the full use of `--enable-framework[=DIR]`, I've tried to bypass this issue by specifying a specific path to install the extra stuff too.

`env PYTHON_CONFIGURE_OPTS="--enable-framework=/Users/techdragon/Library/Frameworks" pyenv install 3.6.3` fails with the same issue.

**Comment by techdragon** [comment]
I've debugged the chain a bit.

1. Not respecting the full version of `--enable-framework[=DIR]` comes from `pyenv` passing its own target version directory path  `~/.pyenv/versions/*` into  `configure` using `--enable-framework[=DIR]` as the final option, which overrides the previous uses of `--enable-framework[=DIR]` in the call to `configure` because final use wins. Understandable behaviour given how pyenv works, one mystery solved.

2. When autoconf builds the makefiles that will compile and install python. While generating the part that handles `--enable-framework`, https://github.com/python/cpython/blob/24c4d046c492d598d32982194c84db236ccf9793/configure.ac#L238, autoconf has "/Application" hardcoded as the path to install the Python Mac Apps into. https://github.com/python/cpython/blob/24c4d046c492d598d32982194c84db236ccf9793/configure.ac#L271 

3. In the same Autoconf section as step 2, autoconf also hardcodes the dependencies of the makefile target `install` to include `frameworkinstallapps` which is the makefile target that performs the installation of the Python Mac Apps into the `/Applications` directory. https://github.com/python/cpython/blob/24c4d046c492d598d32982194c84db236ccf9793/configure.ac#L268

4. We almost have a solution in `python-build` where it handles the `PYTHON_CONFIGURE_OPTS` and other arguments. https://github.com/pyenv/pyenv/blob/7b9d1a40c2e05746330fabef7e3b93e1a552394d/plugins/python-build/bin/python-build#L721 However the logic that handles `PYTHON_MAKE_INSTALL_TARGET` https://github.com/pyenv/pyenv/blob/7b9d1a40c2e05746330fabef7e3b93e1a552394d/plugins/python-build/bin/python-build#L731 and https://github.com/pyenv/pyenv/blob/7b9d1a40c2e05746330fabef7e3b93e1a552394d/plugins/python-build/bin/python-build#L747, together preclude the obvious workaround of telling python to not install the Python Mac Apps, by excluding `frameworkinstallapps` from the make target because `PYTHON_MAKE_INSTALL_TARGET` only accepts a single target string so we can't just pass in all the steps and do ensure_pip after install ourselves. 
```
make: *** No rule to make target `frameworkinstallstructure commoninstall bininstall maninstall frameworkinstallmaclib frameworkinstallunixtools'.  Stop.`
``` 

The real puzzle was that from the look of it, all of this has been unchanged for years and worked fine until now. The `configure.ac` parts involved seem to all be over 8 years old and the parts of `python-build` are over 3 years old. So my best guess is that this is something related to recent versions of macOS having tighter protections on system directories such as the `/Applications` and `/Library` directories. While this is arguably an issue with upstream CPython's autoconf and makefile configuration, an upstream fix would be unlikely to work its way into existing versions of CPython and so those versions would remain broken like this for users of `pyenv`.  

To fix this in `pyenv` there seems to be a few options:
1. Modify `python-build` to support specifying multiple make targets, and document that framework builds on macOS may require specifying `PYTHON_MAKE_INSTALL_TARGET="frameworkinstallstructure commoninstall bininstall maninstall frameworkinstallmaclib frameworkinstallunixtools"` in order to avoid Python trying to install the Python Mac Apps into the hardcoded `/Applications` directory path.
2. Enhance the existing logic around setting `--enable-framework` so that it does one of the following: 
   1. Patch or edit CPython's autoconf files before running configure, so they do not add `frameworkinstallapps` into the final makefiles. 
   2. Patch or edit CPython's autoconf files before running configure, to change the path to somewhere harmless (possibly user configurable).
   3. Patch or edit CPython's makefiles after running configure, so that `frameworkinstallapps` is no longer a dependent make target of the `install` make target.
   4. Patch or edit CPython's makefiles after running configure, to change the path to somewhere harmless (possibly user configurable).

For now I'm locally going to try option 1, since fixing that shortcoming of `PYTHON_MAKE_INSTALL_TARGET` seems somewhat worthwhile regardless of other changes. However it makes sense that for the overall usability of python-build and `pyenv` as a whole, some form of option 2 would also be a good thing to implement. There does appear to be existing tools in `pyenv` that should make this possible. From the look of `setup_builtin_patches` inside `python-build`, it seems that it patches before doing any build steps, so that would make 2.i or 2.ii the easiest options to implement since these might be possible with the existing patch tools and just require determining the broken combinations of CPython and macOS.

**Comment by techdragon** [comment]
I've written a short patch that should work, however it doesn't. For reasons I can't yet fathom, this patch breaks the `bz2`, `readline` and `openssl` modules, despite the install completing correctly and being able to run `-m ensurepip` to update `pip`.

The patch. 
```
--- a/configure
+++ b/configure
@@ -3130,8 +3130,8 @@ if test "${enable_framework+set}" = set; then :
 		PYTHONFRAMEWORKINSTALLDIR=$PYTHONFRAMEWORKPREFIX/$PYTHONFRAMEWORKDIR
 		FRAMEWORKINSTALLFIRST="frameworkinstallstructure"
 		FRAMEWORKALTINSTALLFIRST="frameworkinstallstructure "
-		FRAMEWORKINSTALLLAST="frameworkinstallmaclib frameworkinstallapps frameworkinstallunixtools"
-		FRAMEWORKALTINSTALLLAST="frameworkinstallmaclib frameworkinstallapps frameworkaltinstallunixtools"
+		FRAMEWORKINSTALLLAST="frameworkinstallmaclib frameworkinstallunixtools"
+		FRAMEWORKALTINSTALLLAST="frameworkinstallmaclib frameworkaltinstallunixtools"
 		FRAMEWORKPYTHONW="frameworkpythonw"
 		FRAMEWORKINSTALLAPPSPREFIX="/Applications"
 ```

The output (and for good measure a normal install of 3.6.3 to confirm it was definitely either the framework build itself or something related to my patch that is behind the broken modules). 
```
techdragon$ env PYTHON_CONFIGURE_OPTS="--enable-framework" pyenv install --patch 3.6.3 < ~/dev/pyenv-patch.diff
Installing Python-3.6.3...
patching file configure
patch unexpectedly ends in middle of line
WARNING: The Python bz2 extension was not compiled. Missing the bzip2 lib?
WARNING: The Python readline extension was not compiled. Missing the GNU readline lib?
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems


BUILD FAILED (OS X 10.12.6 using python-build 1.1.5-8-g7b9d1a4)

Inspect or clean up the working tree at /var/folders/pg/hv1fbzjj0dlfkfjk116mz6nr0000gn/T/python-build.20171022182053.85628
Results logged to /var/folders/pg/hv1fbzjj0dlfkfjk116mz6nr0000gn/T/python-build.20171022182053.85628.log

Last 10 log lines:
			upgrade) ensurepip="--upgrade" ;; \
			install|*) ensurepip="" ;; \
		esac; \
		DYLD_FRAMEWORK_PATH=/var/folders/pg/hv1fbzjj0dlfkfjk116mz6nr0000gn/T/python-build.20171022182053.85628/Python-3.6.3 ./python.exe -E -m ensurepip \
			$ensurepip --root=/ ; \
	fi
Collecting setuptools
Collecting pip
Installing collected packages: setuptools, pip
Successfully installed pip-9.0.1 setuptools-28.8.0
techdragon$ pyenv install 3.6.3
Installing Python-3.6.3...
Installed Python-3.6.3 to /Users/techdragon/.pyenv/versions/3.6.3

techdragon$
```

I've attached the full log file output, which as best as I can tell, looks like it has found, and tried to use `readline`, `bz2`, and `openssl` without major issues, until it came time to import them.

[python-build.20171022182053.85628.log](https://github.com/pyenv/pyenv/files/1404794/python-build.20171022182053.85628.log)

**Comment by rcarmo** [comment]
I'm also stumped by this. 

```
PYTHON_CONFIGURE_OPTS="--enable-framework=$(pyenv root)/versions/2.7.15/" pyenv install 2.7.15
```

...does not work and tries to write into `/Applications`, failing with permission denied as per the build log file:

```
...
clang -g -O2 -fno-strict-aliasing -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -o doscript.o -c ./doscript.m
clang -g -O2 -fno-strict-aliasing -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -o main.o -c ./main.m
clang -L/usr/local/opt/readline/lib -L/usr/local/opt/readline/lib -L/usr/local/opt/openssl/lib -L/Users/rcarmo/.pyenv/versions/2.7.15/lib  -o "Python Launcher" FileSettings.o MyAppDelegate.o MyDocument.o PreferencesWindowController.o doscript.o main.o -framework AppKit -framework Carbon
rm -fr "Python Launcher.app"
DYLD_FRAMEWORK_PATH=/var/folders/58/htdqsjc14l16_gpymyhd92l40000gq/T/python-build.20180830112234.46967/Python-2.7.15 ../../python.exe ./../../Lib/plat-mac/bundlebuilder.py \
		--builddir=. \
		--name="Python Launcher" \
		--executable="Python Launcher" \
		--iconfile=./../Icons/PythonLauncher.icns \
		--bundle-id=org.python.PythonLauncher \
		--resource=./../Icons/PythonSource.icns \
		--resource=./../Icons/PythonCompiled.icns \
		--resource=./English.lproj \
		--resource=./factorySettings.plist \
		--plist Info.plist \
		build
Building './Python Launcher.app'
Copying files
Done.
test -d "/Applications/Python 2.7" || mkdir -p "/Applications/Python 2.7"
mkdir: /Applications/Python 2.7: Permission denied
make[2]: *** [install] Error 1
make[1]: *** [install_PythonLauncher] Error 2
make: *** [frameworkinstallapps] Error 2
```

Since I need a `.dylib` version to build PyObj-C apps and don't want to go the kludgey way and run the build through `sudo`, this is a blocking issue for me with `pyenv`...

**Comment by rcarmo** [comment]
I've been digging into the `configure` scripts and the location of the launcher is determined by `FRAMEWORKINSTALLAPPSPREFIX`, which defaults to `/Applications` inside `configure` and does not seem to be trivially overridable.

**Comment by causey** [comment]
Hello!  I'm seeing this issue with Python 2.7.14 - has anyone found a workaround?

**Comment by marekyggdrasil** [comment]
I don't know if it helps anyone, but I found that for some versions it passes, for some it fails, for example this one tried to override framework:

```
env PYTHON_CONFIGURE_OPTS="--enable-framework=$(pyenv root)/versions/3.5.4 CC=clang --enable-unicode --with-threads" pyenv install 3.5.4 -v
```

and this one didn't try to override framework and succeeded

```
env PYTHON_CONFIGURE_OPTS="--enable-framework=$(pyenv root)/versions/3.5.4 CC=clang --enable-unicode --with-threads" pyenv install 3.6.7 -v
```

I think it is maybe because I already had Python 3.5.x installed by default and somehow it was conflicting, can't say for sure...

**Comment by exequiel09** [comment]
Any progress in this. cant install python via pyenv when using non admin user in macOS since it install idle and python apps in /Applications folder

**Comment by phyordia** [comment]
Still no solution 3 years later?

**Comment by bryanmobrien** [comment]
I know this is an old thread, but I got here from a google search.  This isn't a pyenv issue if I understand things correctly.  See [here](https://github.com/pyenv/pyenv/issues/512).   
My workaround was to change the owner (to me) for the /Applications/Python 3.7 directory 
When I ran 

`env PYTHON_CONFIG_OPTS="--enable-framework" pyenv install -v 3.7.8`  

I would get permission denied errors because I didn't have write permissions to the /Applications/Python 3.7 directory.

```
cd /Applications
sudo chown -R yourusername Python\ 3.7
```
I thought this would resolve the issue, so I tried the same command with version 3.8.3.   
I got no errors.  There was no attempt to write to /Applications/Python 3.8.  Weird.  

I think this may have something to do with the fact that I'm relatively new to Python and had also installed Python 3.7 from python.org (not pyenv).

HTH
